graph, graphql: introduce GraphQL spec-compliant `validation` phase and rules

dotansimha opened the pull request:
## Background

While @lutter worked on a refactor for the GraphQL execution (https://github.com/graphprotocol/graph-node/pull/3005), we needed a way to make sure some aspects of execution are valid after merging the refactor PR. 

We also noticed that the current implementation of the GraphQL engine does not include a validation phase (https://github.com/graphprotocol/graph-node/issues/3013). 

This PR uses `graphq-tools` crate (https://github.com/dotansimha/graphql-tools-rs, a set of utils on top of `graphql_parser` crate) for validations, using visitor pattern (similar to GraphQL-JS ecosystem). 

The following rules are now running: 

- `UniqueOperationNames` (https://spec.graphql.org/draft/#sec-Operation-Name-Uniqueness)
- `SingleFieldSubscriptions` (https://spec.graphql.org/draft/#sec-Operation-Name-Uniqueness)
- `KnownTypeNames` (https://spec.graphql.org/draft/#sec-Fragment-Spread-Type-Existence)
- `VariablesAreInputTypes` (https://spec.graphql.org/draft/#sec-Variables-Are-Input-Types)
- `FieldsOnCorrectType` (https://spec.graphql.org/draft/#sec-Field-Selections)
- `LoneAnonymousOperation` (https://spec.graphql.org/draft/#sec-Lone-Anonymous-Operation)
- `FragmentsOnCompositeTypes` (https://spec.graphql.org/draft/#sec-Fragments-On-Composite-Types)
- `OverlappingFieldsCanBeMerged` (https://spec.graphql.org/draft/#sec-Field-Selection-Merging)
- `NoUnusedFragments` (https://spec.graphql.org/draft/#sec-Fragments-Must-Be-Used)
- `KnownFragmentNamesRule` (https://spec.graphql.org/draft/#sec-Fragment-spread-target-defined)
- `LeafFieldSelections` (https://spec.graphql.org/draft/#sec-Leaf-Field-Selections)
- `UniqueFragmentNames` (https://spec.graphql.org/draft/#sec-Fragment-Name-Uniqueness)
- `NoFragmentsCycle` (https://spec.graphql.org/draft/#sec-Fragment-spreads-must-not-form-cycles)
- `PossibleFragmentSpreads` (https://spec.graphql.org/draft/#sec-Fragment-spread-is-possible)
- `NoUnusedVariables` (https://spec.graphql.org/draft/#sec-All-Variables-Used)
- `NoUndefinedVariables` (https://spec.graphql.org/draft/#sec-All-Variable-Uses-Defined)
- `KnownArgumentNames` (https://spec.graphql.org/draft/#sec-Argument-Names , https://spec.graphql.org/draft/#sec-Directives-Are-In-Valid-Locations)
- `UniqueArgumentNames` (https://spec.graphql.org/draft/#sec-Argument-Names)
- `UniqueVariableNames` (https://spec.graphql.org/draft/#sec-Variable-Uniqueness)
- `ProvidedRequiredArguments` (https://spec.graphql.org/draft/#sec-Required-Arguments)

The following spec rules are not part of this PR:
- `ExecutableDefinitions` (not actually needed, because of `graphql_parser` nature)
- `ValuesOfCorrectType` (WIP - not sure yet about requirement by `graph-node`)
- `VariablesInAllowedPosition` (WIP - might be needed by `graph-node`)
- `UniqueInputFieldNames`  (blocked by https://github.com/graphql-rust/graphql-parser/issues/59)
- `KnownDirectives` - not needed at the moment by `graph-node`.
- `UniqueDirectivesPerLocation` - not needed at the moment by `graph-node`.

## Changes in the PR

A `ValidationPlan` struct is now constructed in a static way, including all the rules, and can be disabled using the env var `DISABLE_GRAPHQL_VALIDATIONS=true`.

If there are any validation rules selected, the query fails with a standard error, based on the GraphQL spec:

![image](https://user-images.githubusercontent.com/3680083/146011658-9e0d47a4-4f52-43e4-9619-24053c10a8b4.png)

![image](https://user-images.githubusercontent.com/3680083/146011747-32c15f96-b0b1-4243-9992-946bd1d31570.png)

![image](https://user-images.githubusercontent.com/3680083/146013128-1588a1ee-69c1-47d7-99d0-2883804fc28e.png)

![image](https://user-images.githubusercontent.com/3680083/146034667-73718b8e-392e-4be3-97f7-90789ce0075b.png)

![image](https://user-images.githubusercontent.com/3680083/146034915-85946a53-b08a-4451-a2b2-fd5046f79f61.png)


## Future tasks

- We can consider moving complexity checks and a few more validations into an actual `ValidationRule` and use it there, it should make the GraphQL layer a bit more generic. 
- Drop `validate_fields` since it's covered by new validations. 

## TODO

- [x] Implement `validate` flow based on the spec
- [x] Support `graphql_parser` `Document` 
- [x] Implement initial implementation of `OverlappingFieldsCanBeMerged` 
- [x] Improve `OverlappingFieldsCanBeMerged` implementation
- [x] Add more crucial validation rules

lutter commented on graphql/src/runner.rs:
```diff
@@ -135,12 +141,21 @@ where
+        graphql_validation_plan.add_rule(Box::new(OverlappingFieldsCanBeMerged {}));
+        graphql_validation_plan.add_rule(Box::new(KnownFragmentNamesRule {}));
+        graphql_validation_plan.add_rule(Box::new(NoUnusedFragments {}));
+
```
It would be better IMHO to move this and the call to validating the plan below to `graphql::execution::Query` either as part of `query.validate_fields` [here](https://github.com/graphprotocol/graph-node/blob/master/graphql/src/execution/query.rs#L379) or as a new method `validate_query` on that struct.

dotansimha replied:
I think this will change the nature of `validate` phase to be Query-specific, no? (some validation rules are useful for mutations, and some for subscriptions, so the `validate` phase is for every kind of operation, not only Query...)

I agree that we should find a better place to run it, I was thinking about doing it in `node/src/manager/commands/query.rs` in-between `parse` and `execute`, but we don't have the  `schema::Document` on that point. 

Also, I guess we can remove the `validate_fields` at some point, since it will be executed as validation rule (KnownFields), so it will also run on a separate layer from `Query` ?

lutter replied:
> I think this will change the nature of `validate` phase to be Query-specific, no? (some validation rules are useful for mutations, and some for subscriptions, so the `validate` phase is for every kind of operation, not only Query...)

The `graphql::execution::query::Query` struct is actually used for both subscriptions and queries, so moving it there wouldn't bypass any logic. As for mutations, they are so far out for this part of the code that I don't think it's worth worrying about them.

> I agree that we should find a better place to run it, I was thinking about doing it in `node/src/manager/commands/query.rs` in-between `parse` and `execute`, but we don't have the `schema::Document` on that point.

The code in `node/src/manager` is only used by `graphman` commands, and not used in the `graph-node` service; putting validation there would have no effect for the service.

> Also, I guess we can remove the `validate_fields` at some point, since it will be executed as validation rule (KnownFields), so it will also run on a separate layer from `Query` ?

Yes, that's right; one important wrinkle is [this comment](https://github.com/graphprotocol/graph-node/blob/master/graphql/src/execution/query.rs#L178-L179): it should be a little clearer, but what that is talking about is mutually recursive fragments which can lead to infinite recursion if you just follow them. I haven't looked at the details with `graphql-tools`, but we have to either move any check that is following fragments after the call to `check_complexity` or make sure that `graphql-tools` checks that itself.

dotansimha replied:
I see, let me check, the spec has a rule for recursive fragments, so maybe we can add that as well, and it will simplify some of the checks `graph-node` needs to to today. 

Regarding `Query`, I see what you mean - I'll move it there :) thanks!

dotansimha replied:
Done ;) thanks @lutter

lutter reviewed (changes requested):
Nice progress! Two things:

1. With this approach, each validation requires a pass over the AST; probably not a big deal right now, but something to keep in mind as the number of validations grows.
2. It should be possible to turn all validations off with an environment variable - that's just so that when we roll this out in the hosted service, we can turn it off in case it breaks users' queries to give them some time to fix their queries. In the long run, that env variable will go away, but it's good to have that as we introduce new behavior.

Also, would it be possible to add a validation for [this](https://spec.graphql.org/draft/#sec-Leaf-Field-Selections) We actually have a bug right now where we run queries that violate this sentence:
```
If selectionType is an interface, union, or object

    The subselection set of that selection must NOT BE empty
```

dotansimha commented:
> Nice progress! Two things:

Thank you for the review @lutter ! 

> 1. With this approach, each validation requires a pass over the AST; probably not a big deal right now, but something to keep in mind as the number of validations grows.

I agree, that's why I also created this issue: https://github.com/graphprotocol/graph-node/issues/3077 . We can start with that and then add caching.

> 2. It should be possible to turn all validations off with an environment variable - that's just so that when we roll this out in the hosted service, we can turn it off in case it breaks users' queries to give them some time to fix their queries. In the long run, that env variable will go away, but it's good to have that as we introduce new behavior.

Sure, a flag like `DISABLE_GRAPHQL_VALIDATIONS` will do the trick here, I'll add that :)

> Also, would it be possible to add a validation for [this](https://spec.graphql.org/draft/#sec-Leaf-Field-Selections) We actually have a bug right now where we run queries that violate this sentence:
> 
> ```
> If selectionType is an interface, union, or object
> 
>     The subselection set of that selection must NOT BE empty
> ```

Sure (issue: https://github.com/dotansimha/graphql-tools-rs/issues/3), I'll work on that now.

dotansimha commented:
Ok so now we have more rules :) 

The `LeafFieldSelections` will now check the selection sets, based on the spec definition for that rule: 

```
{
  "errors": [
    {
      "message": "Field \"purpose\" must not have a selection since type \"String\" has no subfields."
    }
  ]
}
```

```
{
  "errors": [
    {
      "message": "Field \"sender\" of type \"Sender!\" must have a selection of subfields. Did you mean \"sender { ... }\"?"
    }
  ]
}
```

dotansimha commented:
@lutter I updated to latest `graphql-tools` and now we have ~12 validation rules. The rules we have now cover everything that `validate_fields` was doing, so I think we can consider removing it?

lutter commented on graphql/src/runner.rs:
```diff
@@ -135,12 +146,30 @@ where
+            graphql_validation_plan.add_rule(Box::new(KnownFragmentNames {}));
+            graphql_validation_plan.add_rule(Box::new(NoUnusedFragments {}));
+            graphql_validation_plan.add_rule(Box::new(OverlappingFieldsCanBeMerged {}));
+        }
```
To me, the validation plan is an implementation detail of how we do validation, and it might be better to just make that a `lazy_static` in `query.rs` itself. I don't think there's much value in having it as a field in the GraphQL runner.

dotansimha replied:
agree, 100000%.

lutter reviewed (changes requested):
This looks great now! I really like how explicit the error messages are.

One small change to move the validation plan out of the runner, and this is good to merge. That change will get rid of a bunch of small modifications to pass the plan around.

Also, before merging, can you squash this? The PR now has a lot of small commits that reflect more this discussion than the sequence of steps needed to get to this point which in the long term don't make that much sense.

As for removing `validate_fields`: yes, I think we can remove this now. I would like to first roll this PR out so that if we have to disable the validations, we still have whatever `validate_fields` does right now, and then follow it with a PR in a week or so that gets rid of the env var to turn validations off and removes `validate_fields`.

dotansimha commented:
> This looks great now! I really like how explicit the error messages are.

❤️ 

> One small change to move the validation plan out of the runner, and this is good to merge. That change will get rid of a bunch of small modifications to pass the plan around.

Done! it's now `lazy_static` ✅ 

> Also, before merging, can you squash this? The PR now has a lot of small commits that reflect more this discussion than the sequence of steps needed to get to this point which in the long term don't make that much sense.

Squashed ✅ 

> As for removing `validate_fields`: yes, I think we can remove this now. I would like to first roll this PR out so that if we have to disable the validations, we still have whatever `validate_fields` does right now, and then follow it with a PR in a week or so that gets rid of the env var to turn validations off and removes `validate_fields`.

I added `Future tasks` to the description, once it's merged, we can open issues for those and track it there?

lutter commented on graphql/src/execution/query.rs:
```diff
@@ -23,6 +25,41 @@ use crate::{
+                Box::new(KnownFragmentNames {}),
+                Box::new(NoUnusedFragments {}),
+                Box::new(OverlappingFieldsCanBeMerged {}),
+                Box::new(NoFragmentsCycle {}),
```
This check needs to be the first one that runs, otherwise other checks that follow fragments can recurse indefinitely.

Actually, looking at the code in `graphql-tools-rs`, `validate` always runs all the validations in the plan, but if fragments contain cycles, we must not ever do anything that recurses on fragments. One simple fix for now would be to call `validate` _after_ the call to `check_complexity` - that checks for cycles and returns an error.

lutter replied:
With that change, the `NoFragmentsCycle` check is actually redundant and should be removed from the plan.

dotansimha replied:
Fixed, and added integration tests to `graphql-tools` to make sure none of the rules are having infinite loops in any rule. 

The GraphQL validation rules are running all validations and do not short-circuit, since it needs to check all rules before returning a result.

The `QueryVisitor` only visits the tree and doesn't resolve `FragmentSpread` so this is something that each rule can do based on its needs.

dotansimha replied:
Also, see `core/tests/interfaces.rs` changes - it's now capturing the cyclic fragments using the validation phase - so if the validation will have an infinite loop - those tests will fail :)

leoyvens commented on server/index-node/Cargo.toml:
```diff
@@ -13,4 +13,4 @@ graphql-parser = "0.4.0"
 hyper = "0.14"
 serde = "1.0"
-either = "1.6.1"
+either = "1.6.1"
```
What happened to the empty newline, some funkiness from your editor?

dotansimha replied:
Yeah for sure, I'll fix that :)

leoyvens commented on store/test-store/Cargo.toml:
```diff
@@ -15,4 +15,4 @@ graph-store-postgres = { path = "../postgres" }
 hex-literal = "0.3"
 diesel = { version = "1.4.8", features = ["postgres", "serde_json", "numeric", "r2d2"] }
-graph-chain-ethereum = { path = "../../chain/ethereum" }
+graph-chain-ethereum = { path = "../../chain/ethereum" }
```
same

dotansimha replied:
fixed :)

otaviopace commented on graphql/tests/query.rs:
```diff
@@ -932,12 +933,12 @@ fn query_complexity_subscriptions() {
-                QueryExecutionError::TooComplex(1_010_200, _) => (), // Expected
-                _ => panic!("did not catch complexity"),
+            Err(SubscriptionError::GraphQLError(e)) => match &e[0] {
+                QueryExecutionError::TooComplex(3030100, _) => (), // Expected
```
I think it's good to keep the underscores in the number since it's pretty big.

dotansimha replied:
fixed

otaviopace commented on graphql/src/execution/query.rs:
```diff
@@ -115,6 +152,24 @@ impl Query {
+                .into_iter()
+                .map(|e| {
+                    QueryExecutionError::ValidationError(
+                        e.locations.clone().into_iter().nth(0),
```
Perhaps here it's possible to just clone the first item instead of the whole `locations`.

dotansimha replied:
sure, fixed!

lutter commented on graphql/tests/query.rs:
```diff
@@ -895,17 +895,18 @@ fn query_complexity_subscriptions() {
+                      members(first: 100, orderBy: id) {
+                          name
+                      }
+                  }
```
What happened here? Why does adding validations require changes to tests that check the complexity calculation?

dotansimha replied:
The subscription was failing now because of `SingleFieldSubscriptions` rule, since using root introspection fields along with a subscription is not valid. 
To achieve a similar complexity level, I changed the subscription to ask for more schema fields instead of introspection fields.

lutter replied:
Ah .. that makes sense. Thanks for clarifying!

lutter commented:
I just approved this, but then noticed that there are changes to tests that I don't understand why they are necessary. It would be good to remove those changes before merging.

Also, the commit message right now is just a log of squashed commits and is not very helpful. It should be changed to something like:
```
graph, graphql: Properly validate GraphQL queries before executing them

Uses `graphql-tools-rs` for the actual validation. Validation can be turned off by setting
`DISABLE_GRAPHQL_VALIDATIONS=true` in the environment
```

dotansimha commented on core/tests/interfaces.rs:
```diff
@@ -279,9 +279,11 @@ async fn follow_interface_reference_invalid() {
-        QueryError::ExecutionError(QueryExecutionError::UnknownField(_, type_name, field_name)) => {
-            assert_eq!(type_name, "Legged");
-            assert_eq!(field_name, "parent");
+        QueryError::ExecutionError(QueryExecutionError::ValidationError(_, error_message)) => {
```
@lutter this one fails now in a different way because of the validations, the tests flow remains the same.

dotansimha commented on core/tests/interfaces.rs:
```diff
@@ -1424,7 +1426,7 @@ async fn recursive_fragment() {
         .unwrap();
     let data = res.to_result().unwrap_err()[0].to_string();
-    assert_eq!(data, "query has fragment cycle including `FooFrag`");
+    assert_eq!(data, "Cannot spread fragment \"FooFrag\" within itself.");
```
Captured by `NoFragmentCycle` rule

dotansimha commented on core/tests/interfaces.rs:
```diff
@@ -1451,5 +1453,8 @@ async fn recursive_fragment() {
-    assert_eq!(data, "query has fragment cycle including `BarFrag`");
+    assert_eq!(
+        data,
+        "Cannot spread fragment \"BarFrag\" within itself via \"FooFrag\"."
```
Captured by NoFragmentCycle rule

dotansimha commented:
> I just approved this, but then noticed that there are changes to tests that I don't understand why they are necessary. It would be good to remove those changes before merging.

Sure. Please see my comments above - I explained on each test what's the reason for changing it. 

> Also, the commit message right now is just a log of squashed commits and is not very helpful. It should be changed to something like:
> 
> ```
> graph, graphql: Properly validate GraphQL queries before executing them
> 
> Uses `graphql-tools-rs` for the actual validation. Validation can be turned off by setting
> `DISABLE_GRAPHQL_VALIDATIONS=true` in the environment
> ```

✅
